skills/repo.go

```diff
@@ -157,6 +168,191 @@ func (r *Repository) RemoveSkill(snapName, skillName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Ensure that such slot exists
+	slot := r.slots[snapName][slotName]
+	if slot == nil {
+		return fmt.Errorf("cannot remove slot %q from snap %q, no such slot", slotName, snapName)
+	}
+	// Ensure that the slot is not using any skills
+	if len(r.skillUsedBySlot[slot]) > 0 {
+		return fmt.Errorf("cannot remove slot %q from snap %q, it still uses granted skills", slotName, snapName)
+	}
+	delete(r.slots[snapName], slotName)
+	if len(r.slots[snapName]) == 0 {
+		delete(r.slots, snapName)
+	}
+	return nil
+}
+
+// Grant grants the named skill to the named slot of the given snap.
+// The skill and the slot must have the same type.
+func (r *Repository) Grant(skillSnapName, skillName, slotSnapName, slotName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Ensure that such skill exists
+	skill := r.skills[skillSnapName][skillName]
+	if skill == nil {
+		return fmt.Errorf("cannot grant skill %q from snap %q, no such skill", skillName, skillSnapName)
+	}
+	// Ensure that such slot exists
+	slot := r.slots[slotSnapName][slotName]
+	if slot == nil {
+		return fmt.Errorf("cannot grant skill to slot %q from snap %q, no such slot", slotName, slotSnapName)
+	}
+	// Ensure that skill and slot are compatible
+	if slot.Type != skill.Type {
+		return fmt.Errorf("cannot grant skill %q from snap %q to slot %q from snap %q, skill type %q doesn't match slot type %q",
+			skillName, skillSnapName, slotName, slotSnapName, skill.Type, slot.Type)
+	}
+	// Ensure that slot and skill are not connected yet
+	if r.skillUsedBySlot[slot][skill] {
+		return fmt.Errorf("cannot grant skill %q from snap %q to slot %q from snap %q twice",
+			skillName, skillSnapName, slotName, slotSnapName)
+	}
+	// Grant the skill
+	if r.skillUsedBySlot[slot] == nil {
+		r.skillUsedBySlot[slot] = make(map[*Skill]bool)
+	}
+	if r.slotsUsingSkill[skill] == nil {
+		r.slotsUsingSkill[skill] = make(map[*Slot]bool)
+	}
+	r.skillUsedBySlot[slot][skill] = true
+	r.slotsUsingSkill[skill][slot] = true
+	return nil
+}
+
+// Revoke revokes the named skill from the slot of the given snap.
+func (r *Repository) Revoke(skillSnapName, skillName, slotSnapName, slotName string) error {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	// Ensure that such skill exists
+	skill := r.skills[skillSnapName][skillName]
+	if skill == nil {
+		return fmt.Errorf("cannot revoke skill %q from snap %q, no such skill", skillName, skillSnapName)
+	}
+	// Ensure that such slot exists
+	slot := r.slots[slotSnapName][slotName]
+	if slot == nil {
+		return fmt.Errorf("cannot revoke skill from slot %q from snap %q, no such slot", slotName, slotSnapName)
+	}
+	// Ensure that slot and skill are connected
+	if !r.skillUsedBySlot[slot][skill] {
+		return fmt.Errorf("cannot revoke skill %q from snap %q from slot %q from snap %q, it is not granted",
+			skillName, skillSnapName, slotName, slotSnapName)
+	}
+	delete(r.skillUsedBySlot[slot], skill)
+	delete(r.slotsUsingSkill[skill], slot)
```

Comment:
As in skill and slot removals, the base maps themselves should have the keys deleted once the respective value is empty, so they don't grow out of bounds.